patches/tModLoader/Terraria/ModLoader/PlayerLoader.cs

```diff
@@ -1326,4 +1326,24 @@ public static List<Item> GetStartingItems(Player player, IEnumerable<Item> vanil
 			.SelectMany(kv => kv.Value)
 			.ToList();
 	}
+
+	/// <summary>
+	/// The action that gets invoked when the item is consumed
+	/// </summary>
+	/// <param name="item">The item that is consumed for crafting</param>
+	/// <param name="index">The index of the item in the List&lt;Item&gt;</param>
+	public delegate void ActionOnUsedForCrafting(Item item, int index);
+	private delegate List<Item> DelegateFindMaterialsFrom(out ActionOnUsedForCrafting onUsedForCrafting);
+	private static HookList HookAddCraftingMaterials = AddHook<DelegateFindMaterialsFrom>(p => p.AddMaterialsForCrafting);
+
+	public static Dictionary<List<Item>, ActionOnUsedForCrafting> GetModdedCraftingMaterials(Player player)
```

Comment:
This should return `IEnumerable<(IEnuermable<Item> items, ActionOnUsedForCrafting onCraft)>`

You can then use `yield return` inside the `foreach` below and yield a value tuple.

Reply:
Just tried it and got the error: `'foreach' statement cannot operate on enumerators of type 'FilteredArrayEnumerator<ModPlayer>' in 'async' or iterator methods because FilteredArrayEnumerator<ModPlayer>' is a 'ref' struct`. I assume that there's something wrong

Reply:
Hmm, unfortunate, we'll need to be a bit smarter then. I'll whip something up